### DISCUSSION TITLE
Alternative for requireUncached when working in vscode

### DISCUSSION THREAD

**Discussion by wusawo** [AnWeber/httpyac#412] [General]
Hi,
I am wondering if there is an alternative to the removed `requireUncached` to make working with own script-files easier during development. 

In the docs it says 
> [NodeJS](https://nodejs.org/api/modules.html#modules_require_cache) caches all loaded scripts. Since in VS Code the script is executed in the context of the extension, the content of the script is not reloaded. Therefore, the script must be manually removed from the cache

How can I manually remove a script from the cache?
Is there a chance to include packages like [import-fresh](https://www.npmjs.com/package/import-fresh) to improve the situation?

Thanks

**Comment by AnWeber** [comment]
My preferred solution is to simply run the scripts via `httpyac` CLI command (https://httpyac.github.io/guide/scripting.html#how-to-debug-scripts). This also makes it easy to debug them directly. Since a new app is launched each time, the caching problem is not present here.
Alternatively it is also possible to simply restart the extension host of vscode (Command `workbench.action.restartExtensionHost`). However, this would be very costly.

> How can I manually remove a script from the cache?

You just have to implement the logic of requireUncached yourself. You need to use `require` method which triggers the loading of the external script.

```javascript
function requireUncached(id: string)  {
  clearModule(id);
  return require(id);
}
function clearModule(id: string, map = new Map())  {
 const module = require.cache[id];
  if (module) {
    map.set(id, true);
    // Clear children modules
    module.children.forEach(child => {
      if (!map.get(child.id)) {
        requireUncached(child.id, map);
      }
    });
    delete require.cache[id];
  }
```

> Is there a chance to include packages like [import-fresh](https://www.npmjs.com/package/import-fresh) to improve the situation?

I don't have a use case for this function in the program and have too many dependencies already anyway. My suggestion would be that you just add this locally in the package.json of your project. These dependencies can be resolved directly. Alternative would be to [build a separate plugin](https://github.com/httpyac/httpyac.github.io/tree/main/examples/plugins/provideVariablesHook/httpyac-plugin-example) for it. `httpyac` is internally just a [collection of plugins](https://github.com/AnWeber/httpyac/tree/main/src/plugins). It is always possible to add functions without my support. A custom plugin would probably be my recommendation as well if you are very intensive in extending functionality with scripts.

**Comment by wusawo** [comment]
Thanks for outlining the various options.
